### PR TITLE
[SPARK-28646][SQL][FOLLOWUP] Add legacy config for allowing parameterless count

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Add a legacy configuration `spark.sql.legacy.allowParameterlessCount` in case users need the parameterless count.
This is a follow-up for https://github.com/apache/spark/pull/30541. 

### Why are the changes needed?

There can be some users depends on the legacy behavior. We need a legacy flag for it.

### Does this PR introduce _any_ user-facing change?

Yes, adding a legacy flag `spark.sql.legacy.allowParameterlessCount`.

### How was this patch tested?

Unit tests

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @beliefer @cloud-fan @HyukjinKwon

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Count.scala [hunk 1]
```diff
@@ -51,8 +52,11 @@ case class Count(children: Seq[Expression]) extends DeclarativeAggregate {
   override def dataType: DataType = LongType
 
   override def checkInputDataTypes(): TypeCheckResult = {
-    if (children.isEmpty) {
-      TypeCheckResult.TypeCheckFailure(s"$prettyName requires at least one argument.")
+    val conf = SQLConf.get
+    if (children.isEmpty && !conf.getConf(SQLConf.ALLOW_PARAMETERLESS_COUNT)) {
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Count.scala:56]: `if (children.isEmpty && !SQLConf.get.getConf(SQLConf.ALLOW_PARAMETERLESS_COUNT)) {`

**Comment by beliefer** [comment]
LGTM. I left one comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -1551,6 +1551,14 @@ object SQLConf {
         "must be positive.")
       .createWithDefault(100)
 
+  val ALLOW_PARAMETERLESS_COUNT =
+    buildConf("spark.sql.legacy.allowParameterlessCount")
+      .internal()
+      .doc("When true, the SQL function `Count` is allowed to take no parameters.")
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1557]:
```suggestion
      .doc("When true, the SQL function 'count' is allowed to take no parameters.")
```

**Review by HyukjinKwon — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]
LGTM

**Review by beliefer — APPROVED** [review]

**Review by beliefer — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.1.

**Pull request closed without merging**